h2o-core/src/main/java/water/udf/fp/FP.java

```diff
@@ -0,0 +1,76 @@
+package water.udf.fp;
+
+import java.util.*;
+
+/**
+ * Elements of Functional Programming (known as FP) in Java
+ * 
+ * @see <a href="https://en.wikipedia.org/wiki/Functional_programming">Wikipedia</a>
+ * for details
+ */
+public class FP {
+  
+  interface Option<T> extends Iterable<T> {
+     T get();
+  }
+  
+  public final static Option<Object> None = new Option<Object>() {
+    @Override
+    public Iterator<Object> iterator() {
+      return Collections.EMPTY_SET.iterator();
+    }
+    
+    @SuppressWarnings("unchecked")
+    public Option<Object> get() { return this; }
+    
+    @Override public String toString() { return "None"; }
+  };
+  
+  public final static class Some<T> implements Option<T> {
+    private List<T> contents;
```

Comment:
Why are we implementing this with a list? Seems like an unnecessary additional object.